src/rvsdg/egglog_optimizer/fast_analyses.rs

```diff
@@ -167,6 +167,101 @@ fn is_pure_rules() -> Vec<String> {
     res
 }
 
+// Tracks what Exprs/Operands/Bodies an operand contains in its context,
+// similar to Body-contains-*.
+// A very important difference is that it does not look into the inputs/outputs of Theta
+// and outputs of Gamma. Instead, it only looks at their inputs.
+//
+// Because there are so many operand, the user of this analysis need
+// to explicitly demand annotate operands they are interested in.
+fn operand_contains_rules() -> Vec<String> {
```

Comment:
As far as I can tell, you rely on these rules saturating then then use negation to run your rule.
That means that if we miss any cases here, it's unsound. What about vecoperand and operandgroup? I don't see rules for them here.